### DISCUSSION TITLE
Nested Array Of Objects into Downloadable CSV in react

### DISCUSSION THREAD

**Discussion by roniceyemeli** [react-csv/react-csv#333] [Q&A]
Hello, I'm looking for help I'm want my elements in the shop array appear in a column  after exporting my file as a csv, here's my code sample https://codesandbox.io/s/frosty-sun-w24h60?file=/src/App.js

**Comment by KeelMatrix** [comment]
The underlying issue is that a nested array of objects has no single CSV representation, so the export has to pick one shape — and `shop` has to become a scalar before it reaches `CSVLink`, because an array passed straight through is stringified to `[object Object],[object Object]`.

Assuming your data looks like `carts = [{ id, customer, shop: [{ name, qty }] }]`, these are the three shapes that work:

**1. One row per cart, array summarised into one cell** — row count matches what the user sees on screen:

```jsx
const rows = carts.map((c) => ({
  id: c.id,
  customer: c.customer,
  shop: (c.shop ?? []).map((s) => `${s.name} x${s.qty}`).join(" | "),
}));
// [{ id: 1, customer: 'A', shop: 'apple x2 | pear x1' },
//  { id: 2, customer: 'B', shop: '' }]
```

**2. One row per array element, parent fields repeated** — best if the CSV goes into a pivot/group-by, but the row count changes:

```jsx
const rows = carts.flatMap((c) => {
  const { shop, ...parent } = c;
  return (shop ?? []).map((s, i) => ({ ...parent, shop_index: i, ...s }));
});
// [{ id: 1, customer: 'A', shop_index: 0, name: 'apple', qty: 2 },
//  { id: 1, customer: 'A', shop_index: 1, name: 'pear',  qty: 1 }]
```

**3. Wide columns** `shop_0_name`, `shop_0_qty`, `shop_1_name`, ... — stable row count, but the column set follows the longest array:

```jsx
const rows = carts.map((c) => {
  const row = { id: c.id, customer: c.customer };
  (c.shop ?? []).forEach((s, i) => {
    for (const [k, v] of Object.entries(s)) row[`shop_${i}_${k}`] = v;
  });
  return row;
});
```

Then it's just `<CSVLink data={rows} filename="carts.csv">Export</CSVLink>`.

Two practical notes:

- Avoid shape 3 if the array length varies a lot (a few long carts turn into hundreds of mostly-empty columns); shape 1 if a human reads the file, shape 2 if a database or spreadsheet formula does.
- If an identifier can start with zeros (SKUs, order numbers), keep it a string in the source object. If it arrives as a number the zeros are already gone before the exporter sees it, and quoting later won't bring them back.

Disclosure: I work on jsonnorm.com, an open JSON→CSV converter — this "array of objects into columns" question is the one it spends most of its UI on (explode to rows vs keep as indexed columns, with the row multiplication shown up front): https://jsonnorm.com/convert/explode-json-array-to-rows/?src=ghdisc
